tgui/packages/tgui/interfaces/PersonalCooking.tsx

```diff
@@ -0,0 +1,481 @@
+  [Food.Meat]: 'bacon',
+  [Food.Nuts]: 'seedling',
+  [Food.Oranges]: 'apple-alt',
+  [Food.Pineapple]: 'apple-alt',
+  [Food.Raw]: 'drumstick-bite',
+  [Food.Seafood]: 'fish',
+  [Food.Sugar]: 'candy-cane',
+  [Food.Toxic]: 'biohazard',
+  [Food.Vegetables]: 'carrot',
+};
+
+const CATEGORY_ICONS = {
+  'Breads': 'bread-slice',
+  'Burgers': 'burger',
+  'Cakes': 'cake-candles',
+  'Egg-Based Food': 'egg',
+  'Frozen': 'ice-cream',
+  'Lizard Food': 'dragon',
+  'Meats': 'bacon',
+  'Mexican Food': 'pepper-hot',
+  'Misc. Food': 'bowl-food',
+  'Mothic Food': 'shirt',
+  'Pastries': 'cookie',
+  'Pies': 'chart-pie',
+  'Pizzas': 'pizza-slice',
+  'Salads': 'leaf',
+  'Sandwiches': 'hotdog',
+  'Seafood': 'fish',
+  'Soups': 'mug-saucer',
+  'Spaghettis': 'wheat-awn',
+};
+
+type Ingredient = {
+  path: string;
+  name: string;
+  amount: number;
+  is_reagent: BooleanLike;
+};
+
+type Recipe = {
+  ref: string;
+  result: string;
+  name: string;
+  desc: string;
+  reqs: Ingredient[];
+  category: string;
+  foodtypes: string[];
+};
+
+type Diet = {
+  liked_food: string[];
+  disliked_food: string[];
+  toxic_food: string[];
+};
+
+type Data = {
+  // Dynamic
+  busy: BooleanLike;
+  display_compact: BooleanLike;
+  display_craftable_only: BooleanLike;
+  craftability: Record<string, BooleanLike>;
+  // Static
+  recipes: Recipe[];
+  categories: string[];
+  foodtypes: string[];
+  diet: Diet;
+};
+
+export const PersonalCooking = (props, context) => {
+  const { act, data } = useBackend<Data>(context);
+  const [searchText, setSearchText] = useLocalState(context, 'searchText', '');
+  const [activeCategory, setCategory] = useLocalState(
+    context,
+    'category',
+    data.categories[0]
+  );
+  const [activeType, setType] = useLocalState(
+    context,
+    'foodtype',
+    data.foodtypes[0]
+  );
+  const [typeMode, setTypeMode] = useLocalState(context, 'typeMode', false);
+  const searchName = createSearch(searchText, (item: Recipe) => item.name);
+  const craftabilityByRef = data.craftability;
+  let recipes = flow([
+    filter<Recipe>(
+      (recipe) =>
+        // If craftable only is selected, then filter by craftability
+        (!data.display_craftable_only ||
+          Boolean(craftabilityByRef[recipe.ref])) &&
+        // Ignore categories and types when searching
+        (searchText.length > 0 ||
+          // Is type mode and the active type matches
+          (typeMode &&
+            (activeType === '' || recipe.foodtypes?.includes(activeType))) ||
+          // Is category mode and the active categroy matches
+          (!typeMode &&
+            (activeCategory === '' || recipe.category === activeCategory)))
+    ),
+    sortBy<Recipe>((recipe) => [
+      -Number(craftabilityByRef[recipe.ref]),
+      recipe.name,
+    ]),
+  ])(data.recipes);
+  if (searchText.length > 0) {
+    recipes = recipes.filter(searchName);
+  }
+  const categories = data.categories.sort();
+  const foodtypes = data.foodtypes.sort();
```

Comment:
Use a `sort` helper because sort method on array mutates the original array, and it's not good if anyone decides to refactor how ui_data is handled.

Reply:
I don't get it, can you please show an example?

Reply:
There's a sort function in common/collections that does the same thing as the sort method. It takes one argument which is the array that you want to sort.

Reply:
https://cartant.medium.com/be-careful-with-array-mutators-2d5231351cf8

Reply:
```
  const categories = canMake.concat(sort(data.categories) as string[]);
  const foodtypes = canMake.concat(sort(data.foodtypes) as string[]);
```
Is this okay?

Reply:
why not just?
```tsx
const categories = sort(data.categories);
const foodtypes = sort(data.foodtypes);
```

Reply:
And if you really need to join lists, use this syntax, it's a bit more visual.

```tsx
const foodtypes = [...canMake, ...sort(data.foodtypes)];
```

Reply:
It doesn't sort them.
```
  const categories_combined = [...canMake, ...sort(categories)];
  const foodtypes_combined = [...canMake, ...sort(foodtypes)];
```
![image](https://user-images.githubusercontent.com/3625094/204566209-c448b97f-6512-434b-87cf-c6421759e72e.png)

Reply:
well, that's interesting...

Reply:
in common/collections.ts, change:
```diff
- export const sort = sortBy();
+ export const sort = sortBy((x) => x);
```

Reply:
Please let me leave it as it is, it doesn't seem to affect the logic. I don't want to dig into shared components without understanding it well.